libraries/chain/include/eosio/chain/wasm_interface_private.hpp

```diff
@@ -96,13 +101,99 @@ struct eosvmoc_tier {
+      }
+#endif
+
+      void apply( const digest_type& code_hash, const uint8_t& vm_type, const uint8_t& vm_version, apply_context& context ) {
+         bool attempt_tierup = false;
+#ifdef EOSIO_EOS_VM_OC_RUNTIME_ENABLED
+         attempt_tierup = eosvmoc && (eosvmoc_tierup == wasm_interface::vm_oc_enable::oc_all || context.should_use_eos_vm_oc());
+         const bool allow_oc_interrupt = attempt_tierup && context.is_applying_block() && context.trx_context.has_undo();
+         if (attempt_tierup) {
+            const chain::eosvmoc::code_descriptor* cd = nullptr;
+            chain::eosvmoc::code_cache_base::get_cd_failure failure = chain::eosvmoc::code_cache_base::get_cd_failure::temporary;
+            try {
+               // Ideally all validator nodes would switch to using oc before block producer nodes so that validators
+               // are never overwhelmed. Compile whitelisted account contracts first on non-produced blocks. This makes
+               // it more likely that validators will switch to the oc compiled contract before the block producer runs
+               // an action for the contract with oc.
+               chain::eosvmoc::code_cache_async::mode m;
+               m.whitelisted = context.is_eos_vm_oc_whitelisted();
+               m.high_priority = m.whitelisted && context.is_applying_block();
+               m.write_window = context.control.is_write_window();
+               auto timer_pause = fc::make_scoped_exit([&](){
+                  context.trx_context.resume_billing_timer();
+               });
+               context.trx_context.pause_billing_timer();
+               cd = eosvmoc->cc.get_descriptor_for_code(m, code_hash, vm_version, failure);
+            } catch (...) {
+               // swallow errors here, if EOS VM OC has gone in to the weeds we shouldn't bail: continue to try and run baseline
+               // In the future, consider moving bits of EOS VM that can fire exceptions and such out of this call path
+               static bool once_is_enough;
+               if (!once_is_enough)
+                  elog("EOS VM OC has encountered an unexpected failure");
+               once_is_enough = true;
+            }
+            if (cd) {
+               if (!context.is_applying_block()) // read_only_trx_test.py looks for this log statement
+                  tlog("${a} speculatively executing ${h} with eos vm oc", ("a", context.get_receiver())("h", code_hash));
+               eosvmoc->exec->execute(*cd, *eosvmoc->mem, context);
+               return;
+            }
+         }
+#endif
+         auto ex = fc::make_scoped_exit([&]() {
+            eos_vm_oc_compile_interrupt = false;
+            executing_code_hash.store({}); // indicate no longer executing
+         });
+         executing_code_hash.store(code_hash);
+         try {
+            get_instantiated_module(code_hash, vm_type, vm_version, context.trx_context)->apply(context);
+         } catch (const interrupt_exception& e) {
```

Comment:
Is it always true that manually triggering the timer will result in an `interrupt_exception` under all conditions? It feels like this can't be true with the current implementation of  `checktime()`?

Reply:
Maybe I'm not understanding what you are getting at. 
The `main_thread_timer.expire_now()` will result in a `interrupt_exception` if applying a block and the timer has not already been triggered. It should not be possible for the timer to have already fired when applying a block.
The code does need to be updated so that `executing_code_hash` is only set if `allow_oc_interrupt` is `true`; looks like I missed that when moving to using code hash.

Reply:
I think maybe the last sentence resolves the path I was thinking through.